src/language/rule/rule.ts

```diff
@@ -89,6 +89,61 @@ export interface IRule {
     applyWithWalker(walker: RuleWalker): RuleFailure[];
 }
 
+export class Replacement {
+    constructor(private innerStart: number, private innerLength: number, private innerText: string) {
+    }
+
+    get start() {
+        return this.innerStart;
+    }
+
+    get length() {
+        return this.innerLength;
+    }
+
+    get end() {
+        return this.innerStart + this.innerLength;
+    }
+
+    get text() {
+        return this.innerText;
+    }
+
+    public apply(content: string) {
+        return content.substring(0, this.start) + this.text + content.substring(this.start + this.length);
+    }
+}
+
+export class Fix {
+    public static applyAll(content: string, fixes: Fix[]) {
+        // accumulate replacements
+        let replacements: Replacement[] = [];
+        for (const fix of fixes) {
+            replacements = replacements.concat(fix.replacements);
+        }
+        // sort in reverse so that diffs are properly applied
+        replacements.sort((a, b) => b.end - a.end);
+        return replacements.reduce((text, r) => r.apply(text), content);
+    }
+
+    constructor(private innerRuleName: string, private innerReplacements: Replacement[]) {
+    }
+
+    get ruleName() {
+        return this.innerRuleName;
+    }
+
+    get replacements() {
+        return this.innerReplacements;
+    }
+
+    public apply(content: string) {
+        // sort replacements in reverse so that diffs are properly applied
+        this.innerReplacements.sort((a, b) => b.end - a.end);
```

Comment:
I'm thinking we could have a `Replacements.applyAll` that is parallel to `Fix.applyAll`?